sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/DisposableMemberInNonDisposableClass.cs

```diff
@@ -28,102 +27,86 @@
-                        Distinct().
-                        Select(symbol => $"'{symbol.Name}'").
-                        OrderBy(s => s));
+                    var message = GetMessage(namedType, analysisContext);
+                    if (string.IsNullOrEmpty(message))
+                    {
+                        return;
+                    }
 
-                    if (!string.IsNullOrEmpty(message))
+                    var typeDeclarations = new CSharpRemovableDeclarationCollector(namedType, analysisContext.Compilation).TypeDeclarations;
+                    foreach (var classDeclaration in typeDeclarations)
                     {
-                        var typeDeclarations = new CSharpRemovableDeclarationCollector(namedType, analysisContext.Compilation).TypeDeclarations;
-                        foreach (var classDeclaration in typeDeclarations)
-                        {
-                            analysisContext.ReportDiagnosticIfNonGenerated(
-                                Diagnostic.Create(rule, classDeclaration.SyntaxNode.Identifier.GetLocation(), message));
-                        }
+                        analysisContext.ReportDiagnosticIfNonGenerated(Diagnostic.Create(rule, classDeclaration.SyntaxNode.Identifier.GetLocation(), message));
                     }
                 },
                 SymbolKind.NamedType);
-        }
 
-        private IEnumerable<IFieldSymbol> GetAssignementsToFieldsIn(IMethodSymbol m, Compilation compilation)
+        private static bool ShouldExclude(ITypeSymbol typeSymbol) =>
+            !typeSymbol.IsClass()
+            || typeSymbol.Implements(KnownType.System_IDisposable)
+            || typeSymbol.Implements(KnownType.System_IAsyncDisposable);
+
+        private static string GetMessage(INamespaceOrTypeSymbol namedType, SymbolAnalysisContext analysisContext)
         {
-            var empty = Enumerable.Empty<IFieldSymbol>();
-            if (m.DeclaringSyntaxReferences.Length != 1)
-            {
-                return empty;
-            }
-            if (m.DeclaringSyntaxReferences[0].GetSyntax() is BaseMethodDeclarationSyntax method)
-            {
-                if (!method.HasBodyOrExpressionBody())
-                {
-                    return empty;
-                }
-                var semanticModel = compilation.GetSemanticModel(method.SyntaxTree);
-                var methodNodes = method.Body == null ?
-                    method.ExpressionBody().DescendantNodes() :
-                    method.Body.DescendantNodes();
-                return methodNodes.
-                    OfType<AssignmentExpressionSyntax>().
-                    Where(n => n.IsKind(SyntaxKind.SimpleAssignmentExpression) && n.Right is ObjectCreationExpressionSyntax).
-                    Select(n => semanticModel.GetSymbolInfo(n.Left).Symbol).
-                    OfType<IFieldSymbol>();
-            }
-            return empty;
+            var disposableFields = namedType.GetMembers()
```

Comment:
if `disposableFields` is empty, you can return early, right?

that would save some unnecessary `GetAssignmentsToFieldsIn` calls which use the semantic model